### ISSUE TITLE
Add test cases to the project to help streamline, well, testing

### ISSUE THREAD

**Issue by Phate6660** [Phate6660/crusty#8]
A current WIP I have modifies `process_input()` and `ShellCommand::run()`, to make them return a bool for if the command succeeded. Which the test function in `src/test.rs` will test for. [Here](https://gist.github.com/Phate6660/172a32169edcf5bb13559e4766468f96) is a gist containing a diff of the current source files, as well as the `src/test.rs` file so you can see what I did for sure.

Personally, I feel could definitely be done much better.
So I thought I should ping you @TheOnAndOnlyZenomat.
You've contributed a lot, and I'd like your input on this.

The idea is to be able test that:
- builtins (`calc`, `echo`, etc)
- external commands (`grep`, `sed`, etc)
- whatever other shell parts

Are working properly all with a single `cargo test`. 
It should also help alleviate the faulty CI.

**Comment by Phate6660** [comment]
Made a branch for this. Posting a gist was weird lol.

**Comment by TheOnAndOnlyZenomat** [comment]
I wanted to implement tests next, actually, so this is very nice timing. This is why I made the builtins return a string, so we can test them separately (and for piping). It would be cool, if we also could test the pipes, so they would have to return a String with the output of the pipe, which then could be printed.
I would have to look a bit closer into the code, to come up with a nice concept for tests, if you'd like that. Or if you want to implement this on you own, no problem. Then I would sit down and update the documentation and maybe format the readme in markdown, just some general housekeeping, if you don't mind.
Feel free to ping me.

**Comment by Phate6660** [comment]
> This is why I made the builtins return a string, so we can test them separately

Oh I didn't even notice that :sweat_smile:
That's a lot better than I was thinking. I got some ideas for this now.
I have [checks clock]... 2 maybe 2.5 hours before school stuff.
I'm gonna rework what I have for testing.

**Comment by Phate6660** [comment]
Yep! Got it looking much nicer. Gonna implement a couple more tests before remaking the testing branch and pushing.

**Comment by TheOnAndOnlyZenomat** [comment]
Looking good. I think it is good, if we do some housekeeping, before pushing new features. Especially testing is good, so we know if we broke something when implementing something new. You can't possibly test all cases with the new code, so good tests are a big plus.

**Comment by Phate6660** [comment]
Forgot to mention, this is being tracked in PR #12 to anyone following this issue.
